l10n_it_fatturapa/bindings/fatturapa.py

```diff
@@ -0,0 +1,147 @@
+# -*- coding: utf-8 -*-
+
+import logging
+import pyxb.binding
+from odoo.modules.module import get_module_resource
+
+from lxml import etree
+from binding import *  # noqa: F403
+
+_logger = logging.getLogger(__name__)
+_logger.setLevel(logging.DEBUG)
+
+XSD_SCHEMA = 'Schema_del_file_xml_FatturaPA_versione_1.2.1.xsd'
+
+_xsd_schema = get_module_resource('l10n_it_fatturapa', 'bindings', 'xsd',
+                                  XSD_SCHEMA)
+_root = etree.parse(_xsd_schema)
+
+_CreateFromDocument = CreateFromDocument  # noqa: F405
+
+date_types = {}
+datetime_types = {}
+string_types = {}
+
+
+def get_parent_element(e):
+    for ancestor in e.iterancestors():
+        if 'name' in ancestor.attrib:
+            return ancestor
+
+
+def get_type_query(e):
+    return "//*[@type='%s']" % e.attrib['name']
+
+
+def collect_element(target, element, parent=None):
+    if parent is None:
+        parent = get_parent_element(element)
+
+    path = '//%s/%s' % (parent.attrib['name'], element.attrib['name'])
+    mandatory = element.attrib.get('minOccurs') != '0'
+    assert path not in target or target[path] == mandatory
```

Comment:
Non mi è ben chiaro cosa dovrebbe fare questo `assert`, forse sarebbe più trasparente usando degli `if`?

Reply:
è un check di debug, si può anche levare

Reply:
in realtà serve per vedere se non lo sto sostituendo con lo stesso elemento ma differente minOccurs, lo metto in modo diverso

Reply:
Ah ok pensavo fosse una specie di filtro, non sapevo cosa facesse `assert`.

Un altro dubbio che avevo su questa parte è: qui `element` è un elemento ben preciso nel XML, quindi perché non prendi il suo path con `getpath`?
Dapprima pensavo che fosse perché poi più tardi trovi tutti i nodi che hanno il path `//parent_name/element_name`, però riguardandolo adesso, non è detto che tutti quegli elementi siano `mandatory`.
Mi sfugge qualcosa?

Reply:
in verità ho preso solo il parent, non credo sia necessario prendere tutto

Reply:
getpath non so se funziona qui, perché questo è l'xsd io ho da trovare il tag nella fattura invece

Reply:
Quindi comunque ci aspettiamo che venga trovato solo un elemento da `//parent_name/element_name`?

Reply:
oppure tutti quelli trovati saranno comunque `mandatory`?

Reply:
> Quindi comunque ci aspettiamo che venga trovato solo un elemento da `//parent_name/element_name`?

o se ce ne sono più di uno che abbia lo stesso valore per mandatory, cioè se lo sovrascrivo ed è obbligatorio o non obbligatorio in entrambi mi va bene uguale

Reply:
Ma se ce ne fossero due con valori diversi non genererebbe problemi sovrascriverli?
Se ad esempio da `root.xpath('//parent_name/element_name')` risultassero due elementi:

- **e1** obbligatorio
- **e2** facoltativo

Sovrascrivendo, può risultare `target['//parent_name/element_name'] = False`.
Questo rimuoverebbe quindi **e1** che non va rimosso.

Probabilmente nei file XML che stai gestendo questo non succede, ma considerato che in sti file sta arrivando di tutto, meglio premunirsi

Reply:
Guardando l'`xsd` questo non succede, e capendo un po' meglio come funzionano, direi che è difficile che succeda quindi direi che va bene anche così. Se ci saranno problemi vedremo che fare

Reply:
Non succede con questo xsd, se domani arriva un 1.3 potrebbe succedere, l'assert è lì per quello, e ci si accorgerebbe subito